### ISSUE TITLE
Automatic Transfer matching? And importer question

### ISSUE THREAD

**Issue by Nixellion** [firefly-iii/firefly-iii#3891]
Hi! First of all, sorry if this is better asked somewhere else, I was not sure if there's like a forum for firefly for such questions? Reddit maybe?

Anyway, I'm currently working on streamlining my import process, and I'm planning to write my own importer for my bank. And I have a couple questions.

# Transfers

First one is regarding matching Transfers. As some other users pointed out before, some banks, including mine, don't provide enough information with their CSV or other export options to match transfers (basically receiving account id or name). Mine just says something like "Transfer between personal accounts" some local tracking IDs and that's all. 

So I was thinking - what about matching by date(time) and exact amount? If one account has debit maybe with certain description that points that it's a transfer, and another account has credit that also says it's a transfer and their dates and amounts match - that's a transfer.

Could that be integrated into Firefly? If not - if I write such a tool, would it be useful?

Basically I'm trying to decide whether I'd like to include this functionality in my importer for my specific bank, or create a more generic tool that would work directly with firefly, processing existing transactions. Meaning that if developer thinks it'll be worth it, or if there are enough people who would benefit from it, I'll consider making it a separate tool. Otherwise I don't yet see a point in it for myself, it'll be a bit easier to just include it in my importer.

# Importer

Are there any importer guidelines? I was thinking about writing it in Python, are there any requirements that importers must follow? Should they all look and feel like Firefly's existing example importers or is that up to the developer? For example since it's an importer for specific bank I don't think I'll need mapping functionality, as all data is there, and it's not flexible and can be hardcoded (like unit conversions and using account IDs to match, since it uses IDs there's no need for name remapping). I'm not sure if UI is needed for that matter, it could just be a single script (well, one that you run once to get oauth running and maybe  set it up, and another to pull data) with some .yaml config for example. I could later add some UI to it just for fun, but I don't see any practical advantage of having a UI in this particular importer.

Thanks a lot for a great tool. Finally starting to get a hang of it, and I hope that this new importer might bring in some more people to it's userbase once it's working ☺️

Also is there a common place for people to find\share importers?

**Comment by JC5** [comment]
Hi! Thanks for some excellend questions.

> So I was thinking - what about matching by date(time) and exact amount?

The original Firefly III import routine used to do that. If you keep it simple it's easy enough, and don't get tempted to add all kinds of exceptions. The original code is gone but you could use [the search](https://docs.firefly-iii.org/concepts/search) to get the same effect. Search for something like `amount:12.34 source_account_is:"bank name" date:2020-01-01` to get started. As such this feature is already there.

The reason the current importers don't do this kind of fuzzy matching anymore is because it's often enough to have Firefly III check for duplicates using the API. Even though that check is pretty strict, it works often enough for most cases.

> Are there any importer guidelines?

None at all. For the importer, you're entirely free to do what you like. At the end of the day you need to submit JSON over the API and how you do it is up to you. There are several [third party tools](https://docs.firefly-iii.org/other-pages/3rdparty) that use all kinds of ways to interact with the API. 

Thanks for the compliments, and let me know if you need anything!

**Comment by Nixellion** [comment]
Thanks! 

Doing this manually is something I'd like to avoid. 

Does Firefly III only store dates of transactions, without exact time?

How does duplicates check work? It's technically not a duplicate if it's debit\credit on **different accounts**. They will also have different descriptions. Also in theory I is possible to have such coincidence that you received 5$ on one account and spent 5$ on another account, happened in the same day but it's not a transfer. 

My bank's API stores date and time so it should be more reliable to figure out transfers while processing it on custom importer's side I suppose.

Well, I guess I'll whip up some simple CLI importer and set it up as a cron job, and let it do the job of matching transactions. And maybe use Specter if I decide to automate PayPal or other banks

Thanks :) I'll share it on github once complete so maybe someone else will find it useful too.

**Comment by JC5** [comment]
Firefly II stores and handles dates + times, yes. 

But let me know what you run into and what you need :+1:

**Comment by Nixellion** [comment]
Thanks a lot for prompt replies, it's a bit surreal to see developer so active :D 

One more question if I may, noticed that CSV has "Internal reference" that's the same for both accounts when transaction happens, so say I import 2 CSVs for 2 accounts and there's transaction going from one account to another, what do I have to do to make Firefly understand that it's a transfer?

**Comment by JC5** [comment]
Just set `type: transfer` and make sure you submit the correct `source_id` or `source_name` fields. 

https://api-docs.firefly-iii.org/#/transactions/storeTransaction

**Comment by Nixellion** [comment]
Sorry, I meant how I could use CSV\Specter Importer and Rules for this? 

API is straightforward and great, no help required there at the moment :)

**Comment by JC5** [comment]
Mmm, not sure. If the Spectre importer is configured correctly it should link to the right accounts but that's a bit of a non-answer. You'll have to do some experimentation.

The rule engine has several triggers and actions that may be useful: change source/destination account, change type of transaction, etc. It's pretty flexible in that regard.

If description is "transfer from A to B" change source to X, dest to Y and set to transfer. Something like that?

**Comment by Nixellion** [comment]
The thing is that there's no "from A to B" or even just "B" in descriptions that my bank provides. I've tried different things by now, and the only way to match transactions like that is by their internal reference or date and amount.

In terms of rules engine I was thinking more along the lines of - it would be cool if it could allow conditions like "If imported transaction's date is equal to another transaction date and amount is equal - do something". But I'm pretty sure that, unless it's already doable, it's gonna be a pain to implement in firefly :(


<RANT START>
god, I've tried everything and why don't banks have common REST APIs? Most don't have them at all, at least in Russia, and from what I've read it's a global problem. But I have 2 different dashboards where I can export 2 differently formatted CSVs neither of which has enough information to determine WHERE the transfer goes. It just says that yeah, it's a transfer to another account. Which one!? 
And why there's no way to automate retrieval of csv or something like that? I ended up using selenium to do it, and got my ip blocked for constant re-logins lol. And obviously their support says "no, something on your end, our bank does not block IPs". Yeah, right. 
</RANT OVER>

**Comment by JC5** [comment]
That's gonna be a pain to introduce into Firefly III, for sure :/

With that little data, perhaps it's an idea to add it with a little script before you feed it to Firefly III.

**Comment by Nixellion** [comment]
Well, yeah, I'm working on that. First trying to just automate import process without using csv importer, and got into an issue. Can't add transaction, I've seen this issue: https://github.com/firefly-iii/firefly-iii/issues/1647 and I think it matches, but I'm still getting an error. Here's the json output with pprint and error message in log.

```
{'apply_rules': 'true',
 'date': '2020-10-04',
 'description': 'test',
 'error_if_duplicate_hash': 'false',
 'transactions': [{'amount': '494.0',
                   'category_name': 'Кафе и рестораны',
                   'date': '2020-10-04',
                   'description': 'test',
                   'internal_reference': 'CRD_7KU30U',
                   'source_name': 'AB Текущий RU счёт',
                   'tags': ['Места-общественного-питания-рестораны'],
                   'type': 'withdrawal'}]}
2020-10-04 23:38:20,812 - default - DEBUG - add_transaction: {'message': 'The given data was invalid.', 'errors': {'transactions.0.description': ['Need at least one transaction.', 'The description field is required.'], 'transactions.0.type': ['Invalid transaction type.']}}
```

Tried like this too:

```
{'apply_rules': 'true',
 'date': '2020-10-04',
 'description': 'test',
 'error_if_duplicate_hash': 'false',
 'transactions': [{'amount': '494.0',
                   'category_name': 'Кафе и рестораны',
                   'internal_reference': 'CRD_7KU30U',
                   'source_name': 'AB Текущий RU счёт',
                   'tags': ['Места-общественного-питания-рестораны']}],
 'type': 'withdrawal'}
2020-10-04 23:40:35,526 - default - DEBUG - add_transaction: {'message': 'The given data was invalid.', 'errors': {'transactions.0.description': ['Need at least one transaction.', 'The description field is required.'], 'transactions.0.type': ['Invalid transaction type.']}}
```

**Comment by Nixellion** [comment]
Ah, found it!

Instead of

```
response = requests.post(self.url + "/api/v1/transactions",
                                 headers=headers,
                                 data=payload)
```

I had to use

```
response = requests.post(self.url + "/api/v1/transactions",
                                 headers=headers,
                                 json=payload)
```

Thanks to @ludo444, found it here: https://gitlab.com/ludo444/fireflyrevoluttransactions/-/blob/master/firefly_revolut_transactions/firefly_revolut_transactions.py

Did not even know that json attribute existed in requests library, it usually works with data. Ah well!

**Comment by Nixellion** [comment]
Well, automation part is taken care of. Ended up using selenium to log into bank website on schedule (cronjob), click through to get to reports page and then parse HTML to get required data for current week. It's better than CSV because it has generalized categories included. I also added MCC codes parsing and adding that data to tags to get a more detailed idea of what kind of merchant that was. 

So it parses data, and adds it to Firefly. For transfers between accounts for now I made it so it adds "Local Transfer - Attention required" tag, and also sends out a Slack alert. So I can go in and manually convert it to transfer with proper account. A better way would be to spend more time in selenium, go through all accounts, get data from all of them, and then match transactions by internal reference codes. Which Firefly could do on it's own actually, now that I think of it, matching part. But for now I think it's easier to just go through and do it manually. It also keeps track of which transactions were already added to firefly with a local sqlite database to not bombard firefly with already added transactions over and over.

And at this point I'm not entirely sure if I should upload it to github. If I do I'll post here, unless there's a better place for the 'announcement'.
